subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,627 @@
+	inst->active_index_subscription.value = BT_GATT_CCC_NOTIFY;
+	atomic_set_bit(inst->active_index_subscription.flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	return bt_gatt_subscribe(inst->conn, &inst->active_index_subscription);
+}
+
+static uint8_t active_index_read_cb(struct bt_conn *conn, uint8_t att_err,
+				    struct bt_gatt_read_params *params, const void *data,
+				    uint16_t len)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	int err = att_err;
+
+	__ASSERT(inst, "no instance for conn %p", (void *)conn);
+
+	BT_DBG("conn %p att_err 0x%02x params %p data %p len %u", (void *)conn, att_err, params,
+	       data, len);
+
+	if (att_err != BT_ATT_ERR_SUCCESS || len == 0) {
+		goto fail;
+	}
+
+	active_index_update(inst, data, len);
+
+	err = active_index_subscribe(inst, params->by_uuid.start_handle);
+	if (err) {
+		BT_ERR("Subscribe failed (err %d)", err);
+		goto fail;
+	}
+
+	return BT_GATT_ITER_STOP;
+
+fail:
+	discover_failed(inst, err);
+	return BT_GATT_ITER_STOP;
+}
+
+static int active_index_read(struct has_inst *inst)
+{
+	BT_DBG("conn %p", (void *)inst->conn);
+
+	(void)memset(&inst->params.read, 0, sizeof(inst->params.read));
+
+	(void)memcpy(&inst->params.uuid, BT_UUID_HAS_ACTIVE_PRESET_INDEX,
+		     sizeof(inst->params.uuid));
+	inst->params.read.func = active_index_read_cb;
+	inst->params.read.handle_count = 0u;
+	inst->params.read.by_uuid.uuid = &inst->params.uuid.uuid;
+	inst->params.read.by_uuid.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	inst->params.read.by_uuid.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+
+	return bt_gatt_read(inst->conn, &inst->params.read);
+}
+
+static void control_point_subscribe_cb(struct bt_conn *conn, uint8_t att_err,
+				       struct bt_gatt_write_params *write)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	int err = att_err;
+
+	__ASSERT(inst, "no instance for conn %p", (void *)conn);
+
+	BT_DBG("conn %p att_err 0x%02x", (void *)inst->conn, att_err);
+
+	if (att_err != BT_ATT_ERR_SUCCESS) {
+		goto fail;
+	}
+
+	err = active_index_read(inst);
+	if (err) {
+		BT_ERR("Active Preset Index read failed (err %d)", err);
+		goto fail;
+	}
+
+	return;
+
+fail:
+	discover_failed(inst, err);
+}
+
+static int control_point_subscribe(struct has_inst *inst, uint16_t value_handle,
+				   uint8_t properties)
+{
+	BT_DBG("conn %p handle 0x%04x", (void *)inst->conn, value_handle);
+
+	inst->control_point_subscription.notify = control_point_notify_cb;
+	inst->control_point_subscription.write = control_point_subscribe_cb;
+	inst->control_point_subscription.value_handle = value_handle;
+	inst->control_point_subscription.ccc_handle = 0x0000;
+	inst->control_point_subscription.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	inst->control_point_subscription.disc_params = &inst->params.discover;
+	if (IS_ENABLED(CONFIG_BT_EATT) && properties & BT_GATT_CHRC_NOTIFY) {
+		inst->control_point_subscription.value = BT_GATT_CCC_INDICATE | BT_GATT_CCC_NOTIFY;
```

Comment:
Why are attempting to write `BT_GATT_CCC_INDICATE` if `properties & BT_GATT_CHRC_NOTIFY`?

Is that unintentional? 

Unless the indicate bit is set, we should not try to subscribe to indications. 

I guess the proper solution here is something like
```c
	if (IS_ENABLED(CONFIG_BT_EATT) && properties & BT_GATT_CHRC_NOTIFY) {
		inst->control_point_subscription.value = BT_GATT_CCC_NOTIFY;
	} 
	
	if (properties & BT_GATT_CHRC_INDICATE) {
		inst->control_point_subscription.value |= BT_GATT_CCC_INDICATE;
	}
```

Reply:
It is intentional. The Indicate property is mandatory while Notify is optional if EATT supported.
The CCCD value is bitmask. So we might want to subscribe for both - indications and notifications. Depending whether EATT channel is open, the ATT Notification (over EATT) or ATT Indication (over ATT) will be sent.

This is how "unreliable" term complicated things.

Reply:
> It is intentional. The Indicate property is mandatory while Notify is optional if EATT supported.

Right, it makes sense to omit the property check if the property is mandatory (although I like defensive programming, and would still perform the check in case of a bad server, but I assume the CCCD write will just fail as an alternative).

Reply:
You're kidding ^^ If so then Read/Write properties should be checked as well. Overkill

Reply:
I was not really kidding :) However in cases like these where it is extremely unlikely that a server is wrongly implemented, it is not necessarily a good idea to perform these checks, even if it could potentially save us a GATT write request :) 

In any case, it is fine as is